pkg/sidecar_controller/snapshot_controller.go

```diff
@@ -0,0 +1,576 @@
+			ctrl.updateContentErrorStatusWithEvent(content, v1.EventTypeWarning, "SnapshotContentCheckandUpdateFailed", fmt.Sprintf("Failed to check and update snapshot content: %v", err))
+			klog.Errorf("checkandUpdateSnapshotStatus [%s]: error occured %v", uniqueKey, err)
+			return err
+		}
+		_, updateErr := ctrl.storeContentUpdate(contentObj)
+		if updateErr != nil {
+			// We will get an "snapshot update" event soon, this is not a big error
+			klog.V(4).Infof("checkandUpdateSnapshotStatus [%s]: cannot update internal cache: %v", uniqueKey, updateErr)
+		}
+
+		return nil
+	})
+	return nil
+}
+
+// updateSnapshotStatusWithEvent saves new snapshot.Status to API server and emits
+// given event on the snapshot. It saves the status and emits the event only when
+// the status has actually changed from the version saved in API server.
+// Parameters:
+//   snapshot - snapshot to update
+//   eventtype, reason, message - event to send, see EventRecorder.Event()
+func (ctrl *csiSnapshotSideCarController) updateContentErrorStatusWithEvent(content *crdv1.VolumeSnapshotContent, eventtype, reason, message string) error {
+	uniqueKey := utils.SnapshotOrContentKey(content)
+	klog.V(5).Infof("updateContentStatusWithEvent[%s]", uniqueKey)
+
+	if content.Status != nil && content.Status.Error != nil && *content.Status.Error.Message == message {
+		klog.V(4).Infof("updateContentStatusWithEvent[%s]: the same error %v is already set", uniqueKey, content.Status.Error)
+		return nil
+	}
+	contentClone := content.DeepCopy()
+	statusError := &crdv1.VolumeSnapshotError{
+		Time: &metav1.Time{
+			Time: time.Now(),
+		},
+		Message: &message,
+	}
+	contentClone.Status.Error = statusError
+	ready := false
+	contentClone.Status.ReadyToUse = &ready
+	newContent, err := ctrl.clientset.SnapshotV1beta1().VolumeSnapshotContents().UpdateStatus(contentClone)
+
+	if err != nil {
+		klog.V(4).Infof("updating VolumeSnapshotContent[%s] error status failed %v", uniqueKey, err)
+		return err
+	}
+
+	_, err = ctrl.storeContentUpdate(newContent)
+	if err != nil {
+		klog.V(4).Infof("updating VolumeSnapshotContent[%s] error status: cannot update internal cache %v", uniqueKey, err)
+		return err
+	}
+	// Emit the event only when the status change happens
+	ctrl.eventRecorder.Event(newContent, eventtype, reason, message)
+
+	return nil
+}
+
+func (ctrl *csiSnapshotSideCarController) getCSISnapshotInput(content *crdv1.VolumeSnapshotContent) (*crdv1.VolumeSnapshotClass, map[string]string, error) {
+	uniqueKey := utils.SnapshotOrContentKey(content)
+	className := content.Spec.VolumeSnapshotClassName
+	klog.V(5).Infof("getCSISnapshotInput for content [%s]: VolumeSnapshotClassName [%s]", content.Name, *className)
+	var class *crdv1.VolumeSnapshotClass
+	var err error
+	if className != nil {
+		class, err = ctrl.getSnapshotClass(*className)
+		if err != nil {
+			klog.Errorf("getCSISnapshotInput failed to getClassFromVolumeSnapshot %s", err)
+			return nil, nil, err
+		}
+	} else {
```

Comment:
how about default snapshot class?

Reply:
That's handled by the common controller.